protocol/lavasession/consumer_session_manager.go

```diff
@@ -224,127 +232,157 @@ func (csm *ConsumerSessionManager) validatePairingListNotEmpty() uint64 {
-				tempIgnoredProviders.providers[providerAddress] = struct{}{}
-			} else if MaximumNumberOfBlockListedSessionsError.Is(err) {
-				// provider has too many block listed sessions. we block it until the next epoch.
-				err = csm.blockProvider(providerAddress, false, sessionEpoch)
-				if err != nil {
-					return nil, 0, "", nil, err
+			// If we successfully got a consumerSession we can apply the current CU to the consumerSessionWithProvider.UsedComputeUnits
+			err = consumerSessionsWithProvider.addUsedComputeUnits(cuNeededForSession)
+			if err != nil {
+				utils.LavaFormatDebug("consumerSessionWithProvider.addUsedComputeUnit", utils.Attribute{Key: "Error", Value: err.Error()})
+				if MaxComputeUnitsExceededError.Is(err) {
+					tempIgnoredProviders.providers[providerAddress] = struct{}{}
+					// We must unlock the consumer session before continuing.
+					consumerSession.lock.Unlock()
+					continue
+				} else {
+					utils.LavaFormatFatal("Unsupported Error", err)
 				}
 			} else {
-				utils.LavaFormatFatal("Unsupported Error", err)
+				// consumer session is locked and valid, we need to set the relayNumber and the relay cu. before returning.
+				consumerSession.LatestRelayCu = cuNeededForSession // set latestRelayCu
+				consumerSession.RelayNum += RelayNumberIncrement   // increase relayNum
+				// Successfully created/got a consumerSession.
+				if debug {
+					utils.LavaFormatDebug("Consumer get session",
+						utils.Attribute{Key: "provider", Value: providerAddress},
+						utils.Attribute{Key: "sessionEpoch", Value: sessionEpoch},
+						utils.Attribute{Key: "consumerSession.CUSum", Value: consumerSession.CuSum},
+						utils.Attribute{Key: "consumerSession.RelayNum", Value: consumerSession.RelayNum},
+						utils.Attribute{Key: "consumerSession.SessionId", Value: consumerSession.SessionId},
+					)
+				}
+				// If no error, add provider session map
+				sessions[providerAddress] = &SessionInfo{
+					Session:           consumerSession,
+					Epoch:             sessionEpoch,
+					ReportedProviders: reportedProviders,
+				}
+
+				// We successfully added provider, we should ignore it if we need to fetch new
+				tempIgnoredProviders.providers[providerAddress] = struct{}{}
+
+				if len(sessions) == wantedSession {
+					return sessions, nil
+				}
+				continue
 			}
-			continue
 		}
 
-		if pairingEpoch != sessionEpoch {
-			// pairingEpoch and SessionEpoch must be the same, we validate them here if they are different we raise an error and continue with pairingEpoch
-			utils.LavaFormatError("sessionEpoch and pairingEpoch mismatch", nil, utils.Attribute{Key: "sessionEpoch", Value: sessionEpoch}, utils.Attribute{Key: "pairingEpoch", Value: pairingEpoch})
-			sessionEpoch = pairingEpoch
+		// If we do not have enough fetch more
+		sessionWithProviderMap, err = csm.getValidConsumerSessionsWithProvider(tempIgnoredProviders, cuNeededForSession, requestedBlock)
```

Comment:
we have discussed getValidConsumerSessionsWithProvider should get an optional argument that specifies how many exactly to expect, this should be a small change in the providerOptimizer code

Reply:
Same as for https://github.com/lavanet/lava/pull/435#discussion_r1208676326